src/test/librbd/test_librbd.cc

```diff
@@ -4477,6 +4477,67 @@ TEST_F(TestLibRBD, TestPendingAio)
   rados_ioctx_destroy(ioctx);
 }
 
+void compare_and_write_copyup(librados::IoCtx &ioctx, bool deep_copyup,
+                              bool *passed)
+{
+  librbd::RBD rbd;
+  std::string parent_name = TestLibRBD::get_temp_image_name();
+  uint64_t size = 2 << 20;
+  int order = 0;
+  ASSERT_EQ(0, create_image_pp(rbd, ioctx, parent_name.c_str(), size, &order));
+
+  librbd::Image parent_image;
+  ASSERT_EQ(0, rbd.open(ioctx, parent_image, parent_name.c_str(), NULL));
+
+  bufferlist bl;
+  bl.append(std::string(4096, '1'));
+  ASSERT_EQ((ssize_t)bl.length(), parent_image.write(0, bl.length(), bl));
+
+  ASSERT_EQ(0, parent_image.snap_create("snap1"));
+  ASSERT_EQ(0, parent_image.snap_protect("snap1"));
+
+  uint64_t features;
+  ASSERT_EQ(0, parent_image.features(&features));
+
+  std::string clone_name = TestLibRBD::get_temp_image_name();
+  EXPECT_EQ(0, rbd.clone(ioctx, parent_name.c_str(), "snap1", ioctx,
+       clone_name.c_str(), features, &order));
+
+  librbd::Image clone_image;
+  ASSERT_EQ(0, rbd.open(ioctx, clone_image, clone_name.c_str(), NULL));
+  if (deep_copyup) {
+    ASSERT_EQ(0, clone_image.snap_create("snap1"));
+  }
+
+  bufferlist write_bl;
+  write_bl.append(std::string(512, '2'));
+  uint64_t mismatch_off;
+  ASSERT_EQ((ssize_t)write_bl.length(),
```

Comment:
@idryomov Seeing a failure here when used against a real OSD [1]

> 2019-02-18T20:24:12.717 INFO:tasks.workunit.client.0.smithi185.stdout:      Expected: (ssize_t)write_bl.length()
> 2019-02-18T20:24:12.717 INFO:tasks.workunit.client.0.smithi185.stdout:      Which is: 512
> 2019-02-18T20:24:12.718 INFO:tasks.workunit.client.0.smithi185.stdout:To be equal to: clone_image.compare_and_write(512, write_bl.length(), bl, write_bl, &mismatch_off, 0)
> 2019-02-18T20:24:12.718 INFO:tasks.workunit.client.0.smithi185.stdout:      Which is: -84
> 

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2019-02-18_14:48:01-rbd-wip-jd-testing-distro-basic-smithi/3607746/teuthology.log

Reply:
This is a bug in compare-and-write itself, working on a fix...

Reply:
Opened https://github.com/ceph/ceph/pull/26519, will rebase on top of it once it's merged.